ir/src/ir2/graph/graph.rs

```diff
@@ -0,0 +1,178 @@
+use crate::graph::{Add, BackLink, Function, Link, NodeTypes};
+use std::{
+    cell::{RefCell, RefMut},
+    fmt::Debug,
+    ops::Deref,
+    rc::Rc,
+};
+
+pub trait Parent: Clone + Into<Link<NodeTypes>> + Debug {
+    fn get_children(&self) -> Link<Vec<Link<NodeTypes>>>;
+    fn add_child(&mut self, child: Link<NodeTypes>) -> Link<NodeTypes> {
+        // This seems to trigger twice when adding a child to a subnode
+        println!(
+            "{:?}.add_child({:?}, {:?})",
+            std::any::type_name::<Self>(),
+            self,
+            child
+        );
+        self.get_children().borrow_mut().push(child.clone());
+        self.clone().into()
+    }
+    fn remove_child(&mut self, child: Link<NodeTypes>) -> Link<NodeTypes> {
+        self.get_children().borrow_mut().retain(|c| c != &child);
+        self.clone().into()
+    }
+    fn first(&self) -> Link<NodeTypes>
+    where
+        Self: Debug,
+    {
+        let children = self.get_children();
+        assert!(
+            !&children.borrow().is_empty(),
+            "first() called on node without children: {:?}",
+            self
+        );
+        let x = children.borrow()[0].clone();
+        x
```

Comment:
A more Rusty way to write it would be:something like:
```suggestion
        self.get_children()
            .borrow()
            .first()
            .expect("first() called on node without children: {:?}", self)
            .clone()
```
(same for last below)

Reply:
Noted and fixed for the next commit on .first()

However, the let x seems necessary for it to compile due to lifetimes being dropped in the wrong order, the fix is also a compiler suggestion

The error was:
```
`children` dropped here while still borrowed... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `Ref<'_, Vec<Link<NodeTypes>>>`
```